metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -1420,6 +1421,58 @@ class MetalsLanguageServer(
     }
   }
 
+  /**
+   * Returns the the definition location or reference locations of a symbol
+   * at a given text document position.
+   * If the symbol represents the definition itself, this method returns
+   * the reference locations, otherwise this returns definition location.
+   * https://github.com/scalameta/metals/issues/755
+   */
+  def definitionOrReferences(
+      position: TextDocumentPositionParams,
+      token: CancelToken = EmptyCancelToken
+  ): Future[DefinitionResult] = {
+    val source = position.getTextDocument.getUri.toAbsolutePath
+    if (source.toLanguage.isScala) {
+      (for {
+        doc <- semanticdbs.textDocument(source).documentIncludingStale
+        positionOccurrence = definitionProvider.positionOccurrence(
+          source,
+          position,
+          doc
+        )
+        occ <- positionOccurrence.occurrence
+      } yield occ) match {
+        case Some(occ) =>
+          if (occ.role.isDefinition) {
+            val referecneContext = new ReferenceContext(false)
+            val refParams = new ReferenceParams(referecneContext)
+            refParams.setTextDocument(position.getTextDocument())
+            refParams.setPosition(position.getPosition())
+            val result = referencesResult(refParams)
+            Future.successful(
+              DefinitionResult(
+                locations = result.locations.asJava,
+                symbol = result.symbol,
+                definition = None,
+                semanticdb = None
+              )
+            )
+          } else {
+            definitionResult(position, token)
+          }
+        case None =>
+          warnings.noSemanticdb(source)
+          // Even if it failed to retrieve the symbol occurrence from semanticdb,
+          // try to find its definitions from presentation compiler.
+          definitionResult(position, token)
```

Comment:
Call `definitionResult` even if it failed to retrieve the symbol information from semanticdb because `DefinitionProvider#definition` [fallback to finding definitions from a presentation compiler](https://github.com/scalameta/metals/blob/50912543e057d958e4a336a6028f74ebca71271e/metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala#L62) if no definition locations found from semanticdb.